firebase-firestore/src/main/java/com/google/firebase/firestore/remote/GrpcCallProvider.java

```diff
@@ -0,0 +1,198 @@
+  private final AsyncQueue asyncQueue;
+
+  private CallOptions callOptions;
+
+  /**
+   * Helper function to globally override the channel that RPCs use. Useful for testing when you
+   * want to bypass SSL certificate checking.
+   *
+   * @param channelBuilderSupplier The supplier for a channel builder that is used to create gRPC
+   *     channels.
+   */
+  @VisibleForTesting
+  public static void overrideChannelBuilder(
+      Supplier<ManagedChannelBuilder<?>> channelBuilderSupplier) {
+    overrideChannelBuilderSupplier = channelBuilderSupplier;
+  }
+
+  GrpcCallProvider(
+      AsyncQueue asyncQueue,
+      Context context,
+      DatabaseInfo databaseInfo,
+      CallCredentials firestoreHeaders) {
+    this.asyncQueue = asyncQueue;
+
+    TaskCompletionSource<ManagedChannel> channelReady = new TaskCompletionSource<>();
+    this.channelTask = channelReady.getTask();
+
+    // We execute network initialization on a separate thread to not block operations that
+    // depend on the AsyncQueue.
+    new Thread("FirestoreGrpcInit") {
+      public void run() {
+        try {
+          ManagedChannel channel = initChannel(context, databaseInfo);
+          FirestoreGrpc.FirestoreStub firestoreStub =
+              FirestoreGrpc.newStub(channel).withCallCredentials(firestoreHeaders);
+          callOptions = firestoreStub.getCallOptions();
+          channelReady.setResult(channel);
+        } catch (Exception e) {
+          channelReady.setException(e);
+        }
+      }
+    }.start();
+  }
+
+  /** Sets up the SSL provider and configures the gRPC channel. */
+  private ManagedChannel initChannel(Context context, DatabaseInfo databaseInfo) {
+    try {
+      // We need to upgrade the Security Provider before any network channels are initialized.
+      // `OkHttp` maintains a list of supported providers that is initialized when the JVM first
+      // resolves the static dependencies of ManagedChannel.
+      ProviderInstaller.installIfNeeded(context);
+    } catch (GooglePlayServicesNotAvailableException /* Thrown by ProviderInstaller */
+        | GooglePlayServicesRepairableException /* Thrown by ProviderInstaller */
+        | IllegalStateException e /* Thrown by Robolectric */) {
+      // Mark the SSL initialization as done, even though we may be using outdated SSL
+      // ciphers. gRPC-Java recommends obtaining updated ciphers from GMSCore, but we allow
+      // the device to fall back to other SSL ciphers if GMSCore is not available.
+      Logger.warn(LOG_TAG, "Failed to update ssl context: %s", e);
+    }
+
+    ManagedChannelBuilder<?> channelBuilder;
+    if (overrideChannelBuilderSupplier != null) {
+      channelBuilder = overrideChannelBuilderSupplier.get();
+    } else {
+      channelBuilder = ManagedChannelBuilder.forTarget(databaseInfo.getHost());
+      if (!databaseInfo.isSslEnabled()) {
+        // Note that the boolean flag does *NOT* indicate whether or not plaintext should be
+        // used
```

Comment:
Any chance you have any idea what this comment means?  It seems nonsensical.

Reply:
I am going to suspect this talks about Protobuf versus Plaintext, so uhm I updated it to the best of my knowledge.